File: files/400-499/431/CH-431.R
library(tidyverse)
library(readxl)

path <- "400-499/431/CH-431 Column Splitting.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "F3:I10")

result <- input %>%
  mutate(
    chars = str_split(ID, ""),
    seen_before = accumulate(lag(chars, default = list(character())), union),
    new_chars = map2(chars, seen_before, ~ unique(.x[!.x %in% .y]))
  ) %>%
  select(ID = new_chars) %>%
  unnest_wider(ID, names_sep = "")

all.equal(result, test)
# True

File: files/400-499/431/CH-431.py
import pandas as pd

path = "400-499/431/CH-431 Column Splitting.xlsx"
input = pd.read_excel(path, usecols="B", nrows=8, skiprows=2)
test = pd.read_excel(path, usecols="F:I", nrows=8, skiprows=2, dtype=str).fillna("")

seen = set()


def new_chars(s):
    global seen
    out = list(dict.fromkeys(c for c in s if c not in seen))
    seen |= set(out)
    return out


result = (
    pd.DataFrame(input["ID"].map(new_chars).tolist())
    .rename(columns=lambda x: f"ID{x + 1}")
    .where(pd.notna, "")
)

print(result.equals(test))
# True
